File: ClipperApp/test_4_windows.cpp
//
// Created by Nail Sharipov on 31.01.2024.
//

#include "test_4_windows.h"
#include "Util.h"
#include "include/clipper2/clipper.h"
#include <iostream>
#include <chrono>
#include <cassert>
#include <iomanip>

using namespace Clipper2Lib;

// A grid of square frames, each with a smaller square cutout in the center.
void WindowsTest::run(int n, ClipType clipType) {
    long long offset = 30;
    long long x = n * offset / 2;
    Point64 origin = {-x, -x};

    auto pair = manyWindows(origin, 20, 10, offset, n);

    int it_count = std::max(500 / n, 1);
    int sq_it_count = it_count * it_count;
    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < sq_it_count; ++i) {
        Paths64 solution = BooleanOp(clipType, FillRule::NonZero, pair.first, pair.second);
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    double time = elapsed.count() / static_cast<double>(sq_it_count);

    int count = 2 * n * n;

    std::cout << count << "     - " << std::fixed << std::setprecision(6) << time << "\n";
}

File: iOverlayAppRust/src/test/test_4_windows.rs
use std::time::Instant;
use i_overlay::core::fill_rule::FillRule;
use i_overlay::core::overlay::Overlay;
use i_overlay::core::overlay_rule::OverlayRule;
use i_overlay::core::solver::Solver;
use i_overlay::i_float::int::point::IntPoint;
use crate::test::util::Util;

pub(crate) struct WindowsTest;
/*
// 4
// Difference:

// multithreading on
8     - 0.000006
32     - 0.000021
128     - 0.000097
512     - 0.000516
2048     - 0.001548
8192     - 0.006780
32768     - 0.034149
131072     - 0.159685
524288     - 0.703147
2097152     - 3.182362
8388608     - 12.058687

// multithreading off
8     - 0.000006
32     - 0.000021
128     - 0.000096
512     - 0.000519
2048     - 0.001675
8192     - 0.007519
32768     - 0.038832
131072     - 0.192338
524288     - 0.835050
2097152     - 3.761808
8388608     - 15.476744
*/

// A grid of square frames, each with a smaller square cutout in the center.
impl WindowsTest {
    pub(crate) fn run(n: usize, rule: OverlayRule, solver: Solver, scale: f64) { // 500
        let offset = 30;
        let x = (n as i32) * offset / 2;
        let origin = IntPoint::new(-x, -x);
        let (subj_paths, clip_paths) = Util::many_windows(origin, 20, 10, offset, n);

        let it_count = ((scale / (n as f64)) as usize).max(1);
        let sq_it_count= it_count * it_count;

        let start = Instant::now();

        for _ in 0..sq_it_count {
            let _ = Overlay::with_contours(&subj_paths, &clip_paths)
                .overlay_with_min_area_and_solver(rule, FillRule::NonZero, 0, solver);
        }

        let duration = start.elapsed();
        let time = duration.as_secs_f64() / sq_it_count as f64;

        let polygons_count = 2 * n * n;

        println!("{}     - {:.6}", polygons_count, time);
    }
}